### PR TITLE
Adding garbage collection for Proof objects.

### PR DESCRIPTION
**Pull request by timothy-king**
Also makes the Proof interface functions const.

Clang-format was applied to a few related locations.

### CODE REVIEW

File: src/proof/proof_manager.cpp [hunk 1]
```diff
@@ -56,43 +56,42 @@ std::string append(const std::string& str, uint64_t num) {
...
-  if (currentPM()->d_fullProof != NULL) {
-    return currentPM()->d_fullProof;
+const Proof& ProofManager::getProof(SmtEngine* smt)
+{
+  if (!currentPM()->d_fullProof)
+  {
+    Assert(currentPM()->d_format == LFSC);
+    currentPM()->d_fullProof.reset(new LFSCProof(
+        smt,
+        static_cast<LFSCCoreSatProof*>(getSatProof()),
```
Comment (4tXJ7f) [inline src/proof/proof_manager.cpp]: We should probably try to get rid of the `static_cast` here but that might be out-of-scope.

File: src/proof/uf_proof.cpp [hunk 2]
```diff
@@ -66,31 +66,37 @@ inline static bool match(TNode n1, TNode n2) {
...
   //AJR : carry this further?
   toStreamLFSC(out, ProofManager::getUfProof(), *d_proof, map);
 }
 
-void ProofUF::toStreamLFSC(std::ostream& out, TheoryProof* tp,
+void ProofUF::toStreamLFSC(std::ostream& out,
+                           TheoryProof* tp,
                            const theory::eq::EqProof& pf,
-                           const ProofLetMap& map) {
+                           const ProofLetMap& map)
```
Comment (4tXJ7f) [inline src/proof/uf_proof.cpp]: Ditto about `const` :)

File: src/proof/arith_proof.cpp [hunk 3]
```diff
@@ -67,26 +67,31 @@ inline static bool match(TNode n1, TNode n2) {
...
+void ProofArith::toStream(std::ostream& out) const
+{
   Trace("theory-proof-debug") << "; Print Arith proof..." << std::endl;
   //AJR : carry this further?
   ProofLetMap map;
   toStreamLFSC(out, ProofManager::getArithProof(), *d_proof, map);
 }
 
-void ProofArith::toStreamLFSC(std::ostream& out, TheoryProof* tp,
+void ProofArith::toStreamLFSC(std::ostream& out,
```
[the remark this answers is not in the document]
Comment (timothy-king) [reply src/proof/arith_proof.cpp]: The function is already static. (Changing the TheoryProof* to const is out of scope.)

File: src/proof/proof_manager.cpp (hunk 1 above)
Comment (timothy-king) [reply src/proof/proof_manager.cpp]: Yes, that is out of scope.

File: src/proof/uf_proof.cpp (hunk 2 above)
Comment (timothy-king) [reply src/proof/uf_proof.cpp]: Ditto about `static` :)

File: src/proof/arith_proof.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (4tXJ7f) [reply src/proof/arith_proof.cpp]: Ha, fair enough :) Sorry, I missed that.

File: src/proof/uf_proof.cpp (hunk 2 above)
Comment (4tXJ7f) [reply src/proof/uf_proof.cpp]: Ditto :)

**Pull request merged**